### ISSUE TITLE
Support for 65c02 mnemonics

### ISSUE THREAD

**Issue by tautology0** [NationalSecurityAgency/ghidra#1261]
**Is your feature request related to a problem? Please describe.**
Yes, it is not possible to fully disassembled 65C02 code currently.

**Describe the solution you'd like**
The 6502 module currently in Ghidra only supports the 6502 set of mnemonics. The 65C02 variant added several more mnemonics (e.g. STZ) which it would be nice to have added.

**Describe alternatives you've considered**
No other alternatives within Ghidra

**Additional context**
Reference to 65C02 mnemonics: https://www.zophar.net/fileuploads/2/10533qqcap/6502ref.html

**Comment by agatti** [comment]
This depends on #1409 being merged/reworked/implemented (BBSx, BBRx, RMBx, and SMBx opcodes need that).

**Comment by tautology0** [comment]
You don't need that change to support 65C02; you could have explicitly set BBS1, BBS2 etc.

**Comment by Godzil** [comment]
The 6502 module is actually weird. It automatically create 2 memory area "STACK" and "LOW_RAM" (Zero Page would be more meaningful) and fill the "low_ram" with label for the Rockwell 6501 (a microcontroller version of the 6502)

Which was fine in my case as I'm analysing a 6501 ROM right now, but there is a catch. The 6501 albeit not being CMOS have the 65C02 instruction set.

Rocwell 6501 datasheet: https://datasheetspdf.com/pdf-file/1340067/Rockwell/R6501/1

I don't understand Guidra why it does thing that all 6502 have these IOs in the zeropage, and not supporing the correct instruction set..

The 6502 shoudl be splitted in 3 variants:
The R6501 with the same memory layout the current 6502 one
A NMOS 6502 with the original 6502 instructions set and maybe no need for the Stack and Low_Ram segments
A 65C02 with the enhance instruction set and like the NMOS 6502 no pre-defined memory segments

**Comment by agatti** [comment]
@Godzil well, actually there's more than just R6501 :)  I wanted to port all the variants I support for https://github.com/agatti/hopper-plugins#6502-cpu-plugin to Ghidra, which means I'd have to do some heavy refactoring of the processor support.  Now, having proper opcode names for 65C02/W65C02/etc. by having something similar to #1409 would be awesome, although @tautology0 is also right in using the "unofficial" bit-based branching opcode names.

The LOW_RAM/ZeroPage renaming is already covered by #1533, actually.

**Comment by tautology0** [comment]
I'd like to see more support for variants, such as for the 65C816 (as I've actually seen that in use on a device that was made this decade and I had to mangle IDA Pro to half disassemble it).

In terms of the labels for zero page I usually just delete LOW_RAM and add my own labels (normally computer specific).

**Comment by agatti** [comment]
In the meantime, I've done some work on this - for now it's at https://github.com/agatti/ghidra/commits/new6502, I'll slowly add support for the ones I already have the definitions for.

**Comment by p2mate** [comment]
Any plans to add support for the 'illegal' opcodes?

**Comment by agatti** [comment]
@p2mate at the moment I don't have much time to add them in my branch - ideally it'd be awesome for somebody to pick that up, finish it, and submit the whole lot as a PR...

**Comment by p2mate** [comment]
@agatti unfortunately my knowledge of sleigh is quite limited :/

**Comment by tautology0** [comment]
@ryanmkurtz thank you!
